lib/Db/SMimeCertificateMapper.php

```diff
@@ -0,0 +1,107 @@
+ *
+ * @license AGPL-3.0-or-later
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\Mail\Db;
+
+use OCA\Mail\Exception\ServiceException;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\Db\MultipleObjectsReturnedException;
+use OCP\AppFramework\Db\QBMapper;
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\IDBConnection;
+
+/**
+ * @template-extends QBMapper<SMimeCertificate>
+ */
+class SMimeCertificateMapper extends QBMapper {
+	/**
+	 * @param IDBConnection $db
+	 */
+	public function __construct(IDBConnection $db) {
+		parent::__construct($db, 'mail_smime_certificates');
+	}
+
+	/**
+	 * @param int $id
+	 * @param string $userId
+	 * @return SMimeCertificate
+	 *
+	 * @throws DoesNotExistException
+	 * @throws ServiceException
+	 */
+	public function find(int $id, string $userId): SMimeCertificate {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where($qb->expr()->eq('id', $qb->createNamedParameter($id)))
+			->andWhere($qb->expr()->eq('user_id', $qb->createNamedParameter($userId)));
+
+		try {
+			return $this->findEntity($qb);
+		} catch (MultipleObjectsReturnedException $e) {
+			// Not possible due to uniqueness of id
+			throw new ServiceException("The impossible has happened", 42, $e);
+		}
+	}
+
+	/**
+	 * @param string $userId
+	 * @return SMimeCertificate[]
+	 */
+	public function findAll(string $userId): array {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where(
+				$qb->expr()->eq('user_id', $qb->createNamedParameter($userId, IQueryBuilder::PARAM_STR))
+			);
+
+		return $this->findEntities($qb);
+	}
+
+	/**
+	 * @param string $userId
+	 * @param string $emailAddress
+	 * @return SMimeCertificate
+	 *
+	 * @throws DoesNotExistException
+	 * @throws MultipleObjectsReturnedException
+	 */
+	public function findByEmailAddress(string $userId, string $emailAddress): SMimeCertificate {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where(
+				$qb->expr()->eq('user_id', $qb->createNamedParameter($userId, IQueryBuilder::PARAM_STR))
+			)
+			->andWhere(
+				$qb->expr()->eq('email_address', $qb->createNamedParameter($emailAddress, IQueryBuilder::PARAM_STR))
+			);
+
+		// TODO: Returning muliple objects should not be possible because we should enforce a single
```

Comment:
should we? what if you want to keep your expired cert so that you can still verify/decrypt old email?

Reply:
Good point! I removed the comment.